// @framer/benjamin.logo/code/Logo.tsx
import * as React from "react"
import { Frame, Size, PropertyControls, ControlType } from "framer"

interface Props {
    company: string
    parentSize: Size
    radius: number
}

interface State {
    company: string
    radius: number
}

export class Logo extends React.Component<Partial<Props>, State> {
    static defaultProps = {
        company: "Framer",
        radius: 100,
        width: 64,
        height: 64
    }

    static propertyControls: PropertyControls<Props> = {
        company: { type: ControlType.String, title: "Company" },
        radius: { type: ControlType.Number, min: 0, max: 100, title: "Radius" }
    }

    state = {
        company: Logo.defaultProps.company,
        radius: Logo.defaultProps.radius
    }

    componentDidMount() {
        const { company, radius } = this.props
        this.setState({ company, radius })
    }

    componentWillReceiveProps(props: Props) {
        if (props.company !== this.props.company) {
            this.setState({ company: props.company, radius: props.radius })
        }
    }

    render() {
        const { company, radius } = this.props
        var logoURL = `https://logo.clearbit.com/${company}.com`;
        var domain = "."

        if (company.includes(domain)) {
            logoURL = `https://logo.clearbit.com/${company}`;
        }

        return (
            <img src={logoURL} style={{ ...style, borderRadius: radius }} />
        )
    }
}

const style: React.CSSProperties = {
    position: "absolute",
    left: 0,
    right: 0,
    top: 0,
    bottom: 0,
    width: "100%",
    height: "100%"
}
